aptos-move/mvhashmap/src/lib.rs

```diff
@@ -2,304 +2,126 @@
         }
     }
 
-    /// For processing outputs - removes the BTreeMap from the MVHashMap.
-    pub fn entry_map_for_key(&self, key: &K) -> Option<BTreeMap<TxnIndex, CachePadded<Entry<V>>>> {
-        self.data
-            .remove(key)
-            .map(|(_, v)| v)
-            .map(|p| p.versioned_map)
-    }
-
-    /// Returns the list of keys that had an associated delta entry at any prior point.
-    pub fn aggregator_keys(&self) -> Vec<K> {
-        std::mem::take(&mut self.delta_keys.lock())
-    }
-
-    /// Add a write of versioned data at a specified key. If the entry is overwritten, asserts
-    /// that the new incarnation is strictly higher.
-    pub fn add_write(&self, key: &K, version: Version, data: V) {
-        let (txn_idx, incarnation) = version;
-
-        let mut v = self.data.entry(key.clone()).or_default();
-        let prev_entry = v.versioned_map.insert(
-            txn_idx,
-            CachePadded::new(Entry::new_write_from(FLAG_DONE, incarnation, data)),
-        );
-
-        // Assert that the previous entry for txn_idx, if present, had lower incarnation.
-        assert!(prev_entry.map_or(true, |entry| -> bool {
-            if let EntryCell::Write(i, _) = entry.cell {
-                i < incarnation
-            } else {
-                true
-            }
-        }));
-    }
-
-    /// Add a delta at a specified key.
-    pub fn add_delta(&self, key: &K, txn_idx: usize, delta: DeltaOp) {
-        let mut v = self.data.entry(key.clone()).or_default();
-        v.versioned_map.insert(
-            txn_idx,
-            CachePadded::new(Entry::new_delta_from(FLAG_DONE, delta)),
-        );
-
-        if !v.contains_delta {
-            v.contains_delta = true;
-            self.delta_keys.lock().push(key.clone());
-        }
+    pub fn take(self) -> (VersionedData<K, V>, VersionedCode<K, V, X>) {
+        (self.data, self.code)
     }
 
     /// Mark an entry from transaction 'txn_idx' at access path 'key' as an estimated write
     /// (for future incarnation). Will panic if the entry is not in the data-structure.
     pub fn mark_estimate(&self, key: &K, txn_idx: TxnIndex) {
-        let v = self.data.get(key).expect("Path must exist");
-        v.versioned_map
-            .get(&txn_idx)
-            .expect("Entry by txn must exist")
-            .mark_estimate();
+        match key.module_path() {
+            Some(_) => self.code.mark_estimate(key, txn_idx),
+            None => self.data.mark_estimate(key, txn_idx),
+        }
     }
 
     /// Delete an entry from transaction 'txn_idx' at access path 'key'. Will panic
     /// if the access path has never been written before.
     pub fn delete(&self, key: &K, txn_idx: TxnIndex) {
-        // TODO: investigate logical deletion.
-        let mut v = self.data.get_mut(key).expect("Path must exist");
-        v.versioned_map.remove(&txn_idx);
+        match key.module_path() {
+            Some(_) => self.code.delete(key, txn_idx),
```

Comment:
we probably should return error or assert on module deletion

Reply:
oh, just realized, this isn't module deletion, this is just for deleting an entry when execution aborts.